New UI

mwobensmith opened the pull request:
Here is a VERY basic demo. Please excuse the lack of styling and non-functioning menu. This is mainly to get feedback on how the grid works.

**To view the demo:**

* Install `bower` with your chosen package manager
* In our project's temp folder, run `bower install bower.json`
* You will be prompted twice regarding component versions to install 
** On both occasions, choose the highest number
* Inside this temp folder, run `python -m SimpleHTTPServer 8082`
* Go to http://localhost:8082


**What we have so far:**
* Sortable grid
* Filters too - see text fields in column footer
** Note: current [bug](https://stackoverflow.com/questions/47021137/vaadin-grid-binding-to-dynamic-filter-variable-when-using-dom-repeat), fields are empty, lacking placeholder text
* Watch number of hosts change in menu, i.e. "Results: 99"
* Resizable and draggable columns

Note that there are four columns right now. Open "transform.json" if you would like to turn fields on or off via "default" parameter. Also, there are many properties in our default log file format that are not represented here, and you can add them using the schema that you see.

**Next steps:**
* Default column widths
* Fields tab, in order to add/remove columns
* Run Info tab, to display run meta data
* Fix filter text field [bug](https://stackoverflow.com/questions/47021137/vaadin-grid-binding-to-dynamic-filter-variable-when-using-dom-repeat)

cr commented:
My feedback so far:

* It might make sense to have one search field covering all the columns.
* It's somewhat limiting that you can only search in columns that are visible.
* It's not clear that the the grid view does not display all the data; very confusing.
* When I scroll to the end of the list, then filter it with just a few results, I ham left with an empty view and have to scroll back up again to see the results.
* There are columns that have just a small set of possible values. A drop down option for the search field would be nice.
* It would be nice to be able to click on an entry to inspect its full log.

mwobensmith commented:
WIP, take 2

Complete rewrite. We are now using bootgrid, built on top of jQuery. Much simpler and better suited for our use case. 

Still a prototype, but mostly working. 

The third party libraries for this grid are linked via a CDN. If we can do this and avoid having to host the libraries in our repo, it seems simpler, but I don't know the pros and cons of this.

Please review for functionality, not code quality. If this works for us, we can eventually upgrade the runs index page as well with a similar design.

TODO list:

* Styling! Ignore color scheme for now
* Date formatting 
* Graph panel
* Metadata panel
* Integration with `log` mode and replacement of `htmlreport`
* Page title
* Code cleanup

mwobensmith commented:
Here goes. All feedback welcome.

cr commented:
I am reviewing this as RTL from your side, so changing the title seems appropriate.

cr commented on tlscanary/template/js/log_handler.js:
```diff
@@ -0,0 +1,75 @@
+var final_data = {};
+var transform_rules;
+
+function test(arg)
```
I'll mark this once, but it's all over the code: The formatting convention in the later JS code is

```javascript
function foo() {
    return;
}

function baz() {
    return;
}
```

with the opening curly trailing on the definition line and one empty line between function definitions.

I'd suggest to globally adapt https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style with long lines for any new code.

cr replied:
Also, is this test function still required?

cr commented on tlscanary/template/js/log_handler.js:
```diff
@@ -0,0 +1,75 @@
+	if (defval == undefined) defval = null;
+	prop = prop.split('.');
+	for (var i = 0; i < prop.length; i++) {
+	    if(typeof obj[prop[i]] == 'undefined')
```
nit: whitespace formatting

also ===

cr commented on tlscanary/template/js/log_handler.js:
```diff
@@ -0,0 +1,75 @@
+
+function find_prop (obj, prop, defval)
+{
+	if (defval == undefined) defval = null;
```
Not sure if you are aware of the difference between lax `==` and strict `===`, but this is a place that probably should use strict comparison. Also, the `typeof foo === 'undefined'` you used three lines down is usually the better test for undefined.

It might make sense to revisit all new JS code and use `==` and `===` as intended. Most times you'll want strict comparison.

mwobensmith replied:
Indeed, strict equality/inequality operators make most sense for string comparison. I am making these changes as requested.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+}
+
+function convertMilliseconds(n) {
+  var hours = Math.floor (n/3600000);
```
nit: mix of formating conventions

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+    var red   = Math.sin(frequency*i + 0) * 127 + 128;
+    var green = Math.sin(frequency*i + 2) * 127 + 128;
+    var blue  = Math.sin(frequency*i + 4) * 127 + 128;
+    a.push (RGB2Color(red,green,blue));
```
nit: whitespace

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+  var element = window.document.getElementById("metadata");
+  var html = "";
+  var argv_args = "";
+  for (var i=0;i<meta.argv.length;i++)
```
nit: whitespacing

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+  {
+    args += i + " : " + meta.args[i] + "<br>";
+  }
+  var metaArray = [];
```
This array can be more efficiently intitalized as 

```javascript
const metaArray = [
    ["...", ...],
    ["...", ...],
    ["...", ...],
    ...
];
```

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+  html += "</tr>";
+  html += "</thead>";
+
+  for (var i=0;i<metaArray.length;i++)
```
weird whitespacing

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+function make_table(hosts, columns)
+{
+  // First, add new column for our grid actions
+  columns.push ({
```
nit: whitespacing here and in following lines

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+    for (var j=0;j<currentColumns.length;j++)
+    {
+      var field = row[currentColumns[j]].toString();
+      if ( field.search(searchStr) != -1 )
```
whitespacing

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+  load_transform();
+}
+
+init();
```
add newline at end of file

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -0,0 +1,447 @@
+function load_transform()
+{
+  var xhr = new XMLHttpRequest();
+  xhr.onload = function(arg) {
```
What about error handling?

cr commented on tlscanary/template/js/transform.json:
```diff
@@ -0,0 +1,224 @@
+		"name": "rank",
+		"type": "int",
+		"default": true,
+		"width":"10%"
```
nit: whitespacing here and other places.

cr commented on tlscanary/template/report_template.htm:
```diff
@@ -1,23 +1,52 @@
+  <script src="../../js/report_page.js"></script>
 </body>
-</html>
+</html>
```
add newline at EOF

cr commented on tlscanary/template/report_template.htm:
```diff
@@ -1,23 +1,52 @@
-  <title>PAGE_TITLE</title>
-  <link rel="stylesheet" href="../../css/report_styles.css">
+<head lang="en">
+  <meta charset="UTF-8">
```
nit: empty HTML elements should be closed with `/>`, for example `<br />`.

I would still land this without, though.

cr commented on tlscanary/report.py:
```diff
@@ -2,15 +2,14 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
 # You can obtain one at http://mozilla.org/MPL/2.0/.
 
+import cert
```
This was *precisely* where it was supposed to be: in the second include block that bundles all the tlscanary modules. Please move it back again.

mwobensmith commented on tlscanary/template/js/log_handler.js:
```diff
@@ -0,0 +1,75 @@
+
```
Quick comment: I'm in the middle of integrating your feedback, but realized just now that `log_handler.js` is an obsolete file. It was from an earlier experiment. All of its functionality was rolled into `report_page.js`, and of course I will be making all of your recommended changes there.

mwobensmith commented:
Addressing feedback.

Regarding whitespace issues, I also fixed this in other places not mentioned. 

In addition, I changed function names to all match the preferred camel case. I also moved a few lines of code around and tweaked a few variable names to make it clearer.

I added the Mozilla PL at the head of the JS file. 

And as per comment above, deleted `log_transform.js` as it was not intended to be part of this PR.

If this looks good, the only thing left to do is merge the changes from master, as this branch was created over a month ago and is missing several recent commits.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -359,89 +344,87 @@ function apply_bootgrid()
\ No newline at end of file
+init();
+
+// License header
```
Not sure what this comment line is for. Is it intentional?

mwobensmith replied:
No, was a TODO comment for me that I need to remove. :)

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -1,80 +1,78 @@
   {
     html += "<option value=\"" + columns[i] + "\"";
-    if ( columns[i] == fieldName )
+    if ( columns[i] === fieldName )
```
nit: whitespaces inside brackets

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -1,80 +1,78 @@
   html += "<h3>Field:&nbsp;&nbsp;<select id=\"fieldNames\" name=\"fieldNames\" >";
   var columns = getVisibleColumns();
-  for (var i=0;i<columns.length;i++)
+  for (var i = 0;i < columns.length;i++)
```
Almost there. According to MDN style guide, for loop formatting is

```js
for (...; ...; ...) {
    ...
}
```

Please also check for loops elsewhere.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -1,80 +1,78 @@
-  var c = window.document.getElementById("chart_canvas");
-  var ctx = c.getContext("2d");
-  window.document.myChart = new Chart(ctx).Pie(data, {animation:false});
+function drawChart (data, fieldName) {
```
nit: whitespace before parenthesis

I have seen this elsewhere. Please check function definitions and calls.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -85,98 +83,89 @@ function resizeChartCanvas()
   return '#' + byte2Hex(r) + byte2Hex(g) + byte2Hex(b);
 }
 
-function returnColorArray (n) {
```
nit: whitespace

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -85,98 +83,89 @@ function resizeChartCanvas()
-  var a = [];
-  var frequency = 0.3;
+function returnColorArray(n) {
+  var colorArray = [];
```
I think this can be const, too.

mwobensmith replied:
Mea culpa - you are right. Seems to be fine. I spoke too soon. :)

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -195,44 +184,43 @@ function makeMetaTab(meta)
-  var tabs = ["results", "chart", "metadata"];
-  for (var i=0;i<tabs.length;i++)
+function navigate(tab) {
+  const $nav = $("#nav");
```
I have missed this before, but what is the thinking behind leading variable names with `$`?

mwobensmith replied:
This comes from jQuery, which we get for free because we are using `bootgrid`. It provides easier ways to look up DOM elements. I have used it in a few places when it was simpler than using a pure JS solution. If consistency is important, I could rewrite all DOM lookups to be either pure JS or jQuery.

cr replied:
Alright, thanks for clarifying. All fine if this was intentional.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -241,59 +229,59 @@ function getVisibleColumns()
       html += "data-visible=\"false\" ";
     }
-    if (columns[i].type == "int")
+    if (columns[i].type === "int")
```
I notice a mixture of quoting styles. We should stick to one unless it really makes sense to use the other (`'"foo" "bar"` vs. `"\"foo\" \"bar\""`), and I think in JS – like python – there is actually no difference between hard `'` and soft `"`.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -185,7 +174,7 @@ function makeMetaTab(meta)
   html += "</thead>";
 
-  for (var i=0;i<metaArray.length;i++)
+  for (var i = 0;i < metaArray.length;i++)
```
You might actually want to check if one of the modern array iterators like `Array.forEach`, `for x in` or `for x of` appeal to you here. There are subtle differences in the way they work, but they should all work fine in this case. They all get rid of the unused iterator variable. Should work for a few other of your for loops as well.

mwobensmith replied:
I will try it. My experience with `Array.forEach` is that - at least in previous ECMAScript implementations - order is not guaranteed. In our project, order is important. The docs say it goes in ascending order, so if that's true, then you are right, we should be using it.

cr replied:
Oh, I thought order was only questionable in objects, but Array ordering is strict.

mwobensmith replied:
OK, after trying to rewrite the `for` loops, I now remember why this isn't a good idea. They tend to enumerate unwanted properties. This is expected for a generic object, but for an indexed array, you get the indexed items AND other weird, undesirable junk that shows up due to things about JS that I don't understand. Regardless, I wasn't able to replace the loops for this reason, so I am going to leave this as-is.

cr replied:
Sounds to me like the difference between `for in` and `for of`. `for of` should both conserve sequence and only iterate just the iterables. Have you tried?

Anyway, nothing to lose sleep over. Using the index is fine, too.

cr replied:
(see [MDN on for..of](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of))

mwobensmith replied:
Good to know, and thank you for the additional info. I will keep this in mind.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -8,7 +8,7 @@ function makeHeaderText(meta) {
           + " vs Fx " + meta.base_metadata.app_version + " " + meta.base_metadata.branch;
   const time = meta.run_start_time.split(".")[0].replace("T","-").replace(":","-").replace(":","-");
-  window.document.getElementById("header").innerHTML = "<h3 class=\"header\">" + desc + "<br>" + time + "</h3>";
+  window.document.getElementById("header").innerHTML = "<h3 class='header'>" + desc + "<br>" + time + "</h3>";
```
Oh, are hard quotes even allowed in html? It definitely looks very alien to me. ```'<h3 class="header">'```, perhaps?

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -27,19 +27,18 @@ function makeChartTab(uriList, fieldName) {
 function makeFieldControl(fieldName) {
   var html = "";
-  html += "<h3>Field:&nbsp;&nbsp;<select id=\"fieldNames\" name=\"fieldNames\" >";
+  html += "<h3>Field:&nbsp;&nbsp;<select id='fieldNames' name='fieldNames' >";
```
Same hardquote question here. Also trailing whitespace before the closing bracket.

cr commented on tlscanary/template/js/report_page.js:
```diff
@@ -27,19 +27,18 @@ function makeChartTab(uriList, fieldName) {
-    html += "<option value=\"" + columns[i] + "\"";
-    if ( columns[i] === fieldName )
+  for (var i = 0;i < columns.length;i++) {
+    html += "<option value='" + columns[i] + "'";
```
More hardquote. I won't mark them all, just wanted to raise the question.

mwobensmith replied:
I assumed that either single or double quote would be valid in HTML. They both seem to work fine, so that counts for something, yes? As long as we are consistent, I have no preference.

cr replied:
Unfortunately "it works" is not a usable metric when it comes to HTML validity. Browsers will try to parse almost anything without complaint. I can't find the relevant HTML specs, so I ran some test input through the W3C validator. It accepts both soft and hard quotes (and even unquoted) without warning, so I guess both are according to spec.

So that degrades it to a style issue. All the documentation and 99.9% of the HTML source code out there always uses soft quotes, and that's why at some point in the future I can see myself normalizing all our HTML to soft quotes, as hard quotes in HTML are somewhat of an eye sore to me.

mwobensmith replied:
Thanks for the research, and I'm fine with whatever format we choose. 

I exported the last two runs in the new report format, and they are now live on the official site:

https://tlscanary.mozilla.org/runs/2017-11-16-09-04-06/
https://tlscanary.mozilla.org/runs/2017-11-15-12-59-15/

\o/
